### PR TITLE
Update default minimum Node.js version to 16.0.0

### PR DESCRIPTION
**Pull request by kripken**
This version of node supports wasm-bigint by default, so it will help #19156

This is about 2 years old, and the node website already provides newer versions
both for latest and for LTS. However, other places are behind, for example the
latest Ubuntu LTS (22) has Node 12, so perhaps we should wait on this.

This is possible after https://github.com/emscripten-core/emsdk/pull/829 made
the emsdk install a 15.x version by default, and then
https://github.com/emscripten-core/emsdk/pull/1232 switched to 16.

### CODE REVIEW

**Review by sbc100 — APPROVED** [review]
lgtm % discussions about when to actually land this

**Comment by jeswr** [comment]
Node 15 no longer has active or security support. Why not go straight to Node 16 which is currently in maintenance LTS?

**Comment by kripken** [comment]
@jeswr Good point, yeah, it might be better to do that. This came up on the mailing list too:

https://groups.google.com/g/emscripten-discuss/c/HmCTeDEw6cQ/m/-nYu5HqfCAAJ

It does seem more obvious to go for 16, but that does allow strictly fewer things than going to 15 (and 15 is the minimum we need for the wasm-bigint feature), so it's not obvious to me what is best.

**Comment by jeswr** [comment]
> so it's not obvious to me what is best.

I'd be suggesting 16. Odd versions of node (including 15) are short lived by design and shouldn't really end up being used in any production scenarios - Node 15 has been deprecated for almost 2 years now.

> This is possible after https://github.com/emscripten-core/emsdk/pull/829 made
the emsdk install a 15.x version by default.

I'd be suggesting to upgrade emsdk to 16.x as well.

**Comment by jeswr** [comment]
Just read the mailing list as well and wanted to respond to a particular comment there.

> Since emsdk (2) has already been updated, I'm not sure there is any reason to update again from 15.14.0 to 16.0.0.  I'm not sure what benefit that would have. 

This would actually be quite a beneficial update in our use case in [swipl-wasm](https://github.com/SWI-Prolog/npm-swipl-wasm). I currently have a draft PR open where we are needing to [switch over to node 16](https://github.com/SWI-Prolog/npm-swipl-wasm/pull/143/files#diff-f34da55ca08f1a30591d8b0b3e885bcc678537b2a9a4aadea4f190806b374ddc) for the build step in order to generate a file that will pass our test suite. Using Node 15 to build some of our tests result in `UnhandledRejectionErrors`. I have *not* dug into what the root cause of these errors were when building with Node 15.

**Comment by sbc100** [comment]
> Just read the mailing list as well and wanted to respond to a particular comment there.
> 
> > Since emsdk (2) has already been updated, I'm not sure there is any reason to update again from 15.14.0 to 16.0.0.  I'm not sure what benefit that would have.
> 
> This would actually be quite a beneficial update in our use case in [swipl-wasm](https://github.com/SWI-Prolog/npm-swipl-wasm). I currently have a draft PR open where we are needing to [switch over to node 16](https://github.com/SWI-Prolog/npm-swipl-wasm/pull/143/files#diff-f34da55ca08f1a30591d8b0b3e885bcc678537b2a9a4aadea4f190806b374ddc) for the build step in order to generate a file that will pass our `ctest` test suite. Using Node 15 to build some of our tests result in `UnhandledRejectionErrors`. I have _not_ dug into what the root cause of these errors were when building with Node 15.

Are the errors you are seeing from running the tests (i.e. running the output of emscripten)?   

Remember this issue is just about the minimum node version that we target by default.   You are still free to run the output on a more recent version, such as 16.. or whatever you want.

**Comment by sbc100** [comment]
If there is an issue with `UnhandledRejectionError` when running on node 15 we should fix that regardless of the default we target since users can always target an explictly older version.   `MIN_NODE_VERSION` can be set to  low as 10.19.0, and we do test the output and run some tests with that ancient version of node.

**Comment by sbc100** [comment]
Looks like node 16 is coming to emsdk: https://github.com/emscripten-core/emsdk/pull/1232

**Comment by kripken** [comment]
Updated to Node 16 after https://github.com/emscripten-core/emsdk/pull/1232 landed.

File: ChangeLog.md
Comment (sbc100) [inline ChangeLog.md:23]: `The default minimum Node version was bumped`

Also can you add the word "default" to the PR title?

**Review by sbc100 — APPROVED** [review]

File: ChangeLog.md
Comment (kripken) [reply ChangeLog.md]: Done, but isn't "was bumped" passive voice, which is generally frowned upon in English? I guess just "bump" as in the old text isn't even grammatical...

**Comment by kripken** [comment]
A code size test improved here. While improving it I also made it work with `--rebaseline`.

**Comment by kripken** [comment]
The code size improvement is due to this diff:
```diff
64,69d63
<  var nodeMajor = process.versions.node.split(".")[0];
<  if (nodeMajor < 15) {
<   process.on("unhandledRejection", reason => {
<    throw reason;
<   });
<  }
```

File: ChangeLog.md
Comment (sbc100) [reply ChangeLog.md]: "Bump the default minimum Node version" sounds correct grammatically but that sounds like that text I would write for the PR itself.   i.e. its in the present tense, imperative mood.   e.d. "Rename foo", "Delete X"

For the changelog entry I normally write in the past tense think.   e.g. "foo was renamed", "x was deleted".   So I would expect "was bumped" here... but I could be wrong.   Reading the rest of the changelog it seems to be mostly be in the past.  Should we swith to using the same tense/mood for PR and for the changelog?

Comment (kripken) [reply ChangeLog.md]: I'm not sure. Let's go with your intuition here (changelog in the past, PRs in the present).

**Comment by sbc100** [comment]
> 15.x -> 16.x in the PR description.

And #1232 is the emsdk PR that made it possible

**Review by sbc100 — APPROVED** [review]

**Comment by kripken** [comment]
This also required a test fix, to make a test use `self.build()`. That way the test uses `self.emcc_args` which we populate with the right node flags for emcc automatically.

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
lgtm!   Once this landed we should do a release I think.

**Comment by kripken** [comment]
Ok, I think this is ready to land. We've collected feedback both here and on the [mailing list](https://groups.google.com/g/emscripten-discuss/c/HmCTeDEw6cQ/m/k4hXVuOMCAAJ) and there were no concerns, aside from actually suggesting to go to 16 and not just 15, as we then agreed.

**Pull request merged**
